### PR TITLE
Print valid instances on target memory on mapper OOM

### PR DESCRIPTION
**Pull request by manopapad**
Example program:

```
import cunumeric as cn
a = cn.ones((1024 * 1024,))
b = a + 1
c = a + 2
d = a + 3
```

New output:

```
> LEGATE_TEST=1 CUNUMERIC_TEST=1 legate --cpus 1 --sysmem 50 oom.py
[0 - 7ff8457a2680]    0.000259 {4}{threads}: reservation ('Python-1 proc 1d00000000000003') cannot be satisfied
[0 - 70000e33f000]    1.481409 {5}{cunumeric.mapper}: Mapper cunumeric on Node 0 failed to allocate 8388608 bytes on memory 1e00000000000000 (of kind SYSTEM_MEM: Visible to all processors on a node) for region requirement 1 of Task cunumeric::BinaryOpTask[oom.py:5] (UID 25).
This means Legate was unable to reserve out of its memory pool the full amount required for the above operation. Here are some things to try:
* Make sure your code is not impeding the garbage collection of Legate-backed objects, e.g. by storing references in caches, or creating reference cycles.
* Ask Legate to reserve more space on the above memory, using the appropriate --*mem legate flag.
* Assign less memory to the eager pool, by reducing --eager-alloc-percentage.
* If running on multiple nodes, increase how often distributed garbage collection runs, by reducing LEGATE_FIELD_REUSE_FREQ (default: 32, warning: may incur overhead).
* Adapt your code to reduce temporary storage requirements, e.g. by breaking up larger operations into batches.
* If the previous steps don't help, and you are confident Legate should be able to handle your code's working set, please open an issue on Legate's bug tracker.
[0 - 70000e33f000]    1.481525 {5}{cunumeric.mapper}: At the point when the above operation would run, the above memory is reserving space for the following PhysicalInstances:
[0 - 70000e33f000]    1.481567 {5}{cunumeric.mapper}: Instance[4000000000000001](region=(1,*,1),memory=1e00000000000000,domain=<0>..<1048575>,fields=1000,layout=SoA:X)
[0 - 70000e33f000]    1.481576 {5}{cunumeric.mapper}:   backing a Store allocated at oom.py:2
[0 - 70000e33f000]    1.481592 {5}{cunumeric.mapper}: Instance[4000000000000003](region=(1,*,1),memory=1e00000000000000,domain=<0>..<1048575>,fields=1002,layout=SoA:X)
[0 - 70000e33f000]    1.481598 {5}{cunumeric.mapper}:   backing a Store allocated at oom.py:4
[0 - 70000e33f000]    1.481613 {5}{cunumeric.mapper}: Instance[4000000000000002](region=(1,*,1),memory=1e00000000000000,domain=<0>..<1048575>,fields=1001,layout=SoA:X)
[0 - 70000e33f000]    1.481618 {5}{cunumeric.mapper}:   backing a Store allocated at oom.py:3
[0 - 70000e33f000]    1.481623 {5}{legate}: Legate called abort in /Users/mpapadakis/legate.core/src/core/mapping/base_mapper.cc at line 954 in function report_failed_mapping
Signal 6 received by node 0, process 16055 (thread 70000e33f000) - obtaining backtrace
Signal 6 received by process 16055 (thread 70000e33f000) at: stack trace: 0 frames
```

### CODE REVIEW

**Comment by manopapad** [comment]
Here is what the latest output looks like:

```
(a) iblis:~/a/cunumeric> LEGATE_TEST=1 legate --cpus 1 --sysmem 50 oom.py
[0 - 7f664d881000]    1.689154 {5}{cunumeric.mapper}: Out of memory: Legate failed to allocate 8388608 bytes on memory 1e00000000000000 (of kind SYSTEM_MEM) for region requirement(s) 1 of Task cunumeric::BinaryOpTask[oom.py:40] (UID 41)
[0 - 7f664d881000]    1.689248 {5}{cunumeric.mapper}: corresponding to a Store allocated at oom.py:40
[0 - 7f664d881000]    1.689372 {5}{cunumeric.mapper}: There is not enough space because Legate is reserving 25166224 of the available 52428800 bytes (minus the eager pool allocation) for the following Instances:
[0 - 7f664d881000]    1.689494 {5}{cunumeric.mapper}: Instance[4000000000000002](region=(1,*,1),memory=1e00000000000000,domain=<0,0>..<3,2>,fields=1000,layout=SoA:XY) of size 96
[0 - 7f664d881000]    1.689512 {5}{cunumeric.mapper}: corresponding to a Store allocated at oom.py:12
[0 - 7f664d881000]    1.689534 {5}{cunumeric.mapper}: Instance[4000000000000003](region=(1,*,1),memory=1e00000000000000,domain=<4,0>..<7,2>,fields=1000,layout=SoA:XY) of size 96
[0 - 7f664d881000]    1.689547 {5}{cunumeric.mapper}: corresponding to a Store allocated at oom.py:12
[0 - 7f664d881000]    1.689581 {5}{cunumeric.mapper}: Instance[4000000000000001](region=(2,*,2),memory=1e00000000000000,domain=<0>..<12>,fields=1000,layout=SoA:X) of size 104
[0 - 7f664d881000]    1.689593 {5}{cunumeric.mapper}: corresponding to a Store allocated at oom.py:23
[0 - 7f664d881000]    1.689616 {5}{cunumeric.mapper}: Instance[4000000000000006](region=(3,*,3),memory=1e00000000000000,domain=<0>..<12>,fields=1048577,layout=SoA:X) of size 104
[0 - 7f664d881000]    1.689628 {5}{cunumeric.mapper}: corresponding to a Store allocated at oom.py:24
[0 - 7f664d881000]    1.689647 {5}{cunumeric.mapper}: Instance[4000000000000004](region=(4,*,4),memory=1e00000000000000,domain=<0>..<1048575>,fields=1001,layout=SoA:X) of size 8388608
[0 - 7f664d881000]    1.689658 {5}{cunumeric.mapper}: corresponding to a Store allocated at oom.py:31[j=13]
[0 - 7f664d881000]    1.689676 {5}{cunumeric.mapper}: Instance[4000000000000005](region=(4,*,4),memory=1e00000000000000,domain=<0>..<1048575>,fields=1002,layout=SoA:X) of size 8388608
[0 - 7f664d881000]    1.689687 {5}{cunumeric.mapper}: corresponding to a Store allocated at oom.py:34
[0 - 7f664d881000]    1.689706 {5}{cunumeric.mapper}: Instance[4000000000000007](region=(4,*,4),memory=1e00000000000000,domain=<0>..<1048575>,fields=1003,layout=SoA:X) of size 8388608
[0 - 7f664d881000]    1.689716 {5}{cunumeric.mapper}: corresponding to a Store allocated at oom.py:37
[0 - 7f664d881000]    1.689723 {5}{cunumeric.mapper}: Here are some things to try:
```

for this example file:

```
from legate.core import Annotation, ingest, float64, TiledSplit
import cunumeric as cn
import numpy as np

# test that ingested Stores show up
tile_shape = (4, 3)
colors = (2, 1)
shape = tuple(ci * di for (ci, di) in zip(colors, tile_shape))
def get_buffer(color):
    (a, b) = color
    return memoryview(np.ones(tile_shape))
tab = ingest(
    float64,
    shape,
    colors,
    TiledSplit(tile_shape),
    get_buffer,
)
arr = cn.array(tab)
arr += 1

# test that unbound stores show up
x = cn.arange(13)
y = cn.unique(x)

# not instantiated => not reported
z = cn.ones((1024 * 1024,))

# test that user annotations show up
with Annotation({"j":"13"}):
    a = cn.ones((1024 * 1024,))

# will succeed
b = a + 1

# will succeed
c = a + 2

# will fail
d = a + 3
```

File: legate/core/runtime.py [hunk 1]
```diff
@@ -131,12 +131,18 @@ def update(self, **kwargs: Any) -> None:
...
+    def machine_string(self) -> str:
         pairs = [f"{key},{value}" for key, value in self._entries.items()]
         if self._provenance is not None:
             pairs.append(f"Provenance,{self._provenance}")
         return "|".join(pairs)
 
+    def user_string(self) -> str:
+        pairs = [f"{key}={value}" for key, value in self._entries.items()]
+        return (
+            "<unknown>" if self._provenance is None else self._provenance
```
Comment (jjwilke) [inline legate/core/runtime.py:143]: no right or wrong way, but I find the following more readable:

```
provenance_str = "<unknown>" if self._provenance is None else self._provenance
if pairs:
  return provenance_str + f"[{','.join(pairs)}]"
return provenance_str
```

File: legate/core/runtime.py [hunk 2]
```diff
@@ -131,12 +131,18 @@ def update(self, **kwargs: Any) -> None:
     def remove(self, key: str) -> None:
         del self._entries[key]
 
-    def __repr__(self) -> str:
+    def machine_string(self) -> str:
```
Comment (jjwilke) [inline legate/core/runtime.py:134]: machine_string and user_string have roughly the same information. Why separate methods?

File: src/core/mapping/base_mapper.cc [hunk 3]
```diff
@@ -767,36 +757,75 @@ void BaseMapper::report_failed_mapping(const Legion::Mappable& mappable,
...
+                 << "objects, e.g. by storing references in caches, or creating reference cycles.";
+  logger.error() << "* Ask Legate to reserve more space on the above memory, using the appropriate "
+                 << "--*mem legate flag.";
+  logger.error() << "* Assign less memory to the eager pool, by reducing --eager-alloc-percentage.";
+  logger.error() << "* If running on multiple nodes, increase how often distributed garbage "
+                 << "collection runs, by reducing LEGATE_FIELD_REUSE_FREQ (default: 32, warning: "
+                 << "may incur overhead).";
+  logger.error() << "* Adapt your code to reduce temporary storage requirements, e.g. by breaking "
+                 << "up larger operations into batches.";
+  logger.error() << "* If the previous steps don't help, and you are confident Legate should be "
```
Comment (jjwilke) [inline src/core/mapping/base_mapper.cc:825]: Add link to bug tracker in the error message

File: src/core/mapping/base_mapper.cc [hunk 4]
```diff
@@ -767,36 +757,75 @@ void BaseMapper::report_failed_mapping(const Legion::Mappable& mappable,
...
+                   << inst.get_instance_size();
+    std::set<Legion::FieldID> fields;
+    inst.get_fields(fields);
+    log_field_info(inst.get_field_space(), fields);
+  }
+
+  logger.error() << "Here are some things to try:";
+  logger.error() << "* Make sure your code is not impeding the garbage collection of Legate-backed "
+                 << "objects, e.g. by storing references in caches, or creating reference cycles.";
+  logger.error() << "* Ask Legate to reserve more space on the above memory, using the appropriate "
```
Comment (jjwilke) [inline src/core/mapping/base_mapper.cc:817]: super nit: "memory using the appropriate" with no comma. There's no emphasized pause. Kind of looks like you are listing elements.

File: src/core/mapping/base_mapper.cc [hunk 5]
```diff
@@ -767,36 +757,75 @@ void BaseMapper::report_failed_mapping(const Legion::Mappable& mappable,
...
+                 << "--*mem legate flag.";
+  logger.error() << "* Assign less memory to the eager pool, by reducing --eager-alloc-percentage.";
```
Comment (jjwilke) [inline src/core/mapping/base_mapper.cc:819]: super nit: "eager pool by reducing" with no comma. There's no emphasized pause. Kind of looks like you are listing elements.

File: src/core/mapping/base_mapper.cc (hunk 4 above)
Comment (jjwilke) [reply src/core/mapping/base_mapper.cc:817]: nit: I might write

"using the appropriate legate flag, e.g. --fbmem for GPU memory or --sysmem for CPU host memory"

File: src/core/mapping/base_mapper.cc [hunk 6]
```diff
@@ -676,7 +675,7 @@ bool BaseMapper::map_legate_store(const Legion::Mapping::MapperContext ctx,
     regions = group->get_regions();
   }
 
-  bool created     = false;
+  bool created     = true;
```
Comment (jjwilke) [inline src/core/mapping/base_mapper.cc:678]: why the change? I would think the initial value would be irrelevant?

**Review by jjwilke — APPROVED** [review]
LGTM. Some small optional nits.

File: src/core/mapping/base_mapper.cc (hunk 6 above)
Comment (manopapad) [reply src/core/mapping/base_mapper.cc:678]: This was a "drive-by" bug fix; the default value was wrong for the [case](https://github.com/nv-legate/legate.core/blob/branch-23.09/src/core/mapping/base_mapper.cc#L697) where `policy.allocation == AllocPolicy::MUST_ALLOC`, because in that case the instance always gets created. The default value doesn't matter for the [case](https://github.com/nv-legate/legate.core/blob/branch-23.09/src/core/mapping/base_mapper.cc#L684) where `policy.allocation == AllocPolicy::MAY_ALLOC`, because it will get overwritten anyway by the `find_or_create_physical_instance` call.

File: legate/core/runtime.py (hunk 1 above)
Comment (bryevdv) [reply legate/core/runtime.py:143]: TBH I would probably just live with `[]` in the empty case for simplicity:
```
return (self._provenance or "<unknown>") + + f"[{','.join(pairs)}]"
```

Comment (bryevdv) [reply legate/core/runtime.py:143]: Or if you really want it to look nice in that case:
```python
entries = ",".join(f"{k}={v}" for k, v in self._entries.items())
provenance = self._provenance or "<unknown>"
return provenance + f" [{entries}]" if entries else ""
```

File: legate/settings.py [hunk 7]
```diff
@@ -56,6 +56,17 @@ class LegateRuntimeSettings(Settings):
         """,
     )
 
+    full_bt_on_oom: PrioritizedSetting[bool] = PrioritizedSetting(
```
Comment (bryevdv) [inline legate/settings.py:59]: Any way to avoid two initialisms at once? `full_trace_on_oom` or `full_oom_debug`?

**Comment by manopapad** [comment]
Deprecated by another PR
